[SC-247] Endpoint do dashboardu studenta

barto14753 opened the pull request:
Endpoint w takiej formie jak w opisie na jirze (https://systematic-chaos.atlassian.net/jira/software/projects/SC/boards/1?selectedIssue=SC-247). Na ten moment zamockowełem 3 pola w heroStats związane z rangami, levelami i badgami, których  jeszcze na backu jeszcze nie ma (jeśli się mylę to mnie poprawcie). Co do expa to liczyłem go proporcjonalnie do liczby zdobytych punktów (dla survey i info max za każdym razem)

barto14753 committed:
[SC-247] Added dashboard endpoint skeleton

barto14753 committed:
[SC-247] Added GeneralStats for dashboard endpoint response

barto14753 committed:
[SC-247] Merge branch 'develop' of https://github.com/ajablonsk1/BEng-Thesis into SC-247-endpoint-do-dashboardu-studenta

barto14753 committed:
[SC-247] Added LastAddedActivities for dashboard endpoint response

barto14753 committed:
[SC-247] Added HeroStats for dashboard endpoint response

barto14753 committed:
[SC-247] Changed experience counting

dmncp commented:
Zauważyłem że w lastAddedActivities jest pole availableUntil, które @barto14753 ustalił jako string. To moja wina bo tak napisałem w jirze bo tak miałem w mocku dla ułatwienia. Natomiast chyba w każdym endpoincie podajecie w milisekundach. Więc możecie to zamienić na milisekundy jeżeli tak będzie bardziej spójnie albo czytelniej. Dla frontu to obojętne całkowicie.

Dodatkowo prośba do @ajablonsk1: zweryfikuj czy to available until dobrze współpracuje z tym wymaganiem w praktyce bo ja sprawdziłem i u mnie nie działa. Dodałem wymaganie na datę zamknięcia w wymaganiach a ta zmienna availableUntil dalej byla null. Sprawdź też czy exp liczy się jak należy bo miałeś chyba na to jakiś pomysł liczenia.

dmncp reviewed (changes requested):
Daję blocka bo u mnie nie działa availableUntil

ajablonsk1 commented on api/src/main/java/com/example/api/controller/user/DashboardController.java:
```diff
@@ -0,0 +1,26 @@
+@SecurityRequirement(name = "JWT_AUTH")
+public class DashboardController {
+    private final DashboardService dashboardService;
+    @GetMapping("")
```
("") redundantne

ajablonsk1 commented on api/src/main/java/com/example/api/dto/response/user/dashboard/GeneralStats.java:
```diff
@@ -0,0 +1,17 @@
+@Setter
+public class GeneralStats {
+    private Double avgGraphTask;
+    private Double avgCombatTask;
```
jak nazywamy wszędzie FileTask to tu też nazwij, jak chcesz żeby w jsonie nazywało sie inaczej nazywało to dajesz @JsonProperty("avgCombatTask"), ale wsm nie wiem po co zawsze używaliśmy FileTask po co zmieniać

ajablonsk1 commented on api/src/main/java/com/example/api/service/user/DashboardService.java:
```diff
@@ -0,0 +1,274 @@
+
+    private GeneralStats getGeneralStats(User student) {
+        Double avgGraphTask = getAvgGraphTask(student);
+        Double avgCombatTask = getAvgCombatTask(student);
```
tu też zmień na fileTask

ajablonsk1 replied:
wszędzie niżej zmień na fileTask

ajablonsk1 commented on api/src/main/java/com/example/api/service/user/DashboardService.java:
```diff
@@ -0,0 +1,274 @@
+        );
+    }
+
+    private Double getExperiencePoints(User student) {
```
Ogólnie to jest pole w studencie points (to są własnie punkty doświadczenia - zmieniłem, bo i tak wszędzie używamy określenia punkty a nie doświdaczenie), zostało zrobione po to, żeby po każdej ocenie dodawały sie tam punkty - nie wiem czy to jesy robione teraz, jeżeli nie to zamiast tego (bo to się będzie liczyć za każdym razem jak ktos odświeży strone) albo daj taska na Jire żeby punkty się dodawały do pola points po każdej zakończonej aktywności, albo sam to zrób

barto14753 replied:
To dodawanie punktów po aktywnościach zrobie w oddzielnym tasku, żeby nie mieszać i wykorzystam to nowe pole zamiast liczenia tego za każdym razem

ajablonsk1 commented:
> Zauważyłem że w lastAddedActivities jest pole availableUntil, które @barto14753 ustalił jako string. To moja wina bo tak napisałem w jirze bo tak miałem w mocku dla ułatwienia. Natomiast chyba w każdym endpoincie podajecie w milisekundach. Więc możecie to zamienić na milisekundy jeżeli tak będzie bardziej spójnie albo czytelniej. Dla frontu to obojętne całkowicie.
> 
> Dodatkowo prośba do @ajablonsk1: zweryfikuj czy to available until dobrze współpracuje z tym wymaganiem w praktyce bo ja sprawdziłem i u mnie nie działa. Dodałem wymaganie na datę zamknięcia w wymaganiach a ta zmienna availableUntil dalej byla null. Sprawdź też czy exp liczy się jak należy bo miałeś chyba na to jakiś pomysł liczenia.

Noo wszystkie daty trzymamy w milisekundach, więc @barto14753 zmien, nie wiem czy to naprawi, ale sprawdz potem czy to zadziała, Jeżeli nie to spróbuj naprawić, jak nie będziesz wiedział to daj znać, bo wsm sie nie pytasz o nic, a potem wychodzi tak jak wyżej, że piszesz ogromną funkcję do liczenia punktów, a pole points w userze jest po to - tak jak pisałem nw cz aktualnie jest to updatowane, ale łatwiej updatowć i potem getem brać niż za każdym razem liczyć od nowa. Co do punktów doświadczenia to napsiałem wyżej, i punty doświadczenia niech mapują się 1:1 z punktami zdobytymi, i tak nie nazywamy tego doświadczeniem na froncie i na backu więc bez sensu kombinować

barto14753 committed:
[SC-247] Renamed combatTask to fileTask

barto14753 committed:
[SC-247] Changed availableUntil to Long

barto14753 committed:
[SC-247] Set one activity to return non-null availableUntil in dashboard response

barto14753 commented:
Ustawiłem drafta bo zrobię to naliczanie punktów userowi w oddzielnym  tasku. Co do availableUntil to zmieniłem na Longa a jedno z wymagań w configu odpowiadające za availableUntil ("Date from") ustawiłemu mu selected na true przez co endpoint zwraca jednego non-null'a. @ajablonsk1 @dmncp  Zastanawiam się po co jest ten atrybut selected w wymaganiach i dlaczego w configu wszystkie wymagania miały do tej pory selected=false? Obecnie w serwisie dashboardu filtruje wymagania tak aby miały selected=true, ale może to nie potrzebne.

ajablonsk1 commented:
> Ustawiłem drafta bo zrobię to naliczanie punktów userowi w oddzielnym tasku. Co do availableUntil to zmieniłem na Longa a jedno z wymagań w configu odpowiadające za availableUntil ("Date from") ustawiłemu mu selected na true przez co endpoint zwraca jednego non-null'a. @ajablonsk1 @dmncp Zastanawiam się po co jest ten atrybut selected w wymaganiach i dlaczego w configu wszystkie wymagania miały do tej pory selected=false? Obecnie w serwisie dashboardu filtruje wymagania tak aby miały selected=true, ale może to nie potrzebne.

Selected oznacza, to czy prowadzący chce aby dane wymaganie było brane pod uwagę. Rozpiszę teraz flow tych wymagań, bo może być to niejasne.

Przy tworzeniu jakiejkolwiek aktywności są dodawane do niej defaultowe wymagania, które wyglądaja przykładowo tak:

```
[
    {
        "id": 1,
        "name": <nazwa>,
        "value": null,
        "type": "DATE",
        "selected": false
    },
    {
        "id": 2,
        "name": <nazwa>,
        "value": null,
        "type": "DATE",
        "selected": false
    },
    {
        "id": 3,
        "name": <nazwa>,
        "value": 100.0,
        "type": "NUMBER",
        "selected": false
    },
    {
        "id": 4,
        "name": <nazwa>,
        "value": [],
        "type": "MULTI_SELECT",
        "selected": false
    },
    {
        "id": 5,
        "name": <nazwa>,
        "value": [],
        "type": "MULTI_SELECT",
        "selected": false
    },
    {
        "id": 6,
        "name": <nazwa>,
        "value": [],
        "type": "MULTI_SELECT",
        "selected": false
    },
    {
        "id": 7,
        "name": <nazwa>,
        "value": [],
        "type": "MULTI_SELECT",
        "selected": false
    }
]
```
<nazwa> jest dodana już na pr tym co leży moim, ale na tym co teraz byłem nie było i dlatego tak dałem, ale tam jest string po polsku z nazwą req

Są to defaultowe requirementy. Gdy prowadzący je zmieni - stają się nie defaultowe. Jeżeli prowadzący zmieni requirement i da mu jakieś value, ale selected wciąż będzie false, to nie bedzie brane pod uwagę. To jest spoko jak np prowadzący chce sobie szybko wyłączyć jakies wymaganie to wchodzi zmienia na false zapisuje i już nie jest brane pod uwage to wymaganie.

ajablonsk1 commented:
Czyli podsumowując flow:
1. Prowadzący tworzy aktywność - dodawane są defaultowe wymagania, które blokują wyświetlanie się aktywności dla każdego (defaultowe są takie jak wysłałem wyżej - mają wszędzie na selected false)
2. Dopiero, gdy prowadzący ustawi wymagania (mogą być puste), defaultowe zmieniają się na normalne i są brane pod uwage.
3. Wymagania sprawdzane są tak, że sprawdzamy każde najpierw czy jest selected, jeżeli nie to pomijamy, jeżeli tak to sprawdzamy czy jest spełniane.

Nie wiem czy jest to jasne, jak coś to daj znać

barto14753 committed:
[SC-247] Merge remote-tracking branch 'origin/develop' into SC-247-endpoint-do-dashboardu-studenta

# Conflicts:
#	api/src/main/java/com/example/api/service/map/ChapterService.java
#	api/src/main/java/com/example/api/service/map/RequirementService.java

barto14753 committed:
[SC-247] Adjusted calls of areRequirementsFulfilled to pass student as argument

barto14753 committed:
[SC-247] Made experiencePoints to be taken from User entity

barto14753 committed:
[SC-247] Added saving users (again) to update their points

barto14753 committed:
[SC-247] Took into account in dashboard endpoint only activities with non-default requirements

barto14753 commented:
- zastąpiłem liczenie exp'a użyciem pola points w Userze (to jest 1:1 liczba zdobytych punktów z aktywności i dodatkowych punktów). Ogólnie to dodawanie punktów userowi działa noramalnie jak korzystam z gui, ale do configu musiałem dodać ponowny save userów, bo inaczej nie dodawało tych punktów
- w przypadku lastAddedActivities to biorę od tej pory pod uwagę tylko te które NIE MAJĄ defaultowych requirementów, bo z tego co zrozumiałem aktywności z defaultowymi nie powinny być widoczne dla studentów (@ajablonsk1 popraw mnie jeśli się mylę). Od tej pory powinniście widzieć availableUntil w postaci Long'a. Screen na dowód:

<img width="467" alt="Zrzut ekranu 2022-09-18 o 22 07 33" src="https://user-images.githubusercontent.com/56938330/190926339-a38dc06d-fb30-49ae-b196-49f7b01de704.png">

ajablonsk1 commented:
> * zastąpiłem liczenie exp'a użyciem pola points w Userze (to jest 1:1 liczba zdobytych punktów z aktywności i dodatkowych punktów). Ogólnie to dodawanie punktów userowi działa noramalnie jak korzystam z gui, ale do configu musiałem dodać ponowny save userów, bo inaczej nie dodawało tych punktów
> * w przypadku lastAddedActivities to biorę od tej pory pod uwagę tylko te które NIE MAJĄ defaultowych requirementów, bo z tego co zrozumiałem aktywności z defaultowymi nie powinny być widoczne dla studentów (@ajablonsk1 popraw mnie jeśli się mylę). Od tej pory powinniście widzieć availableUntil w postaci Long'a. Screen na dowód:
> 
> <img alt="Zrzut ekranu 2022-09-18 o 22 07 33" width="467" src="https://user-images.githubusercontent.com/56938330/190926339-a38dc06d-fb30-49ae-b196-49f7b01de704.png">

Tak, jak aktywność ma defaultowe wymagania to tak jakby jej nie było, jest jakby w stanie oczekiwania na prowadzącego aż ustawi wymagania, po czym jak ustawi, to sprawdzamy czy są spełnione.

ajablonsk1 reviewed (approved):
Do kodu brak większych zastrzeżeń, jak działa jak powinno to git

barto14753 committed:
Merge branch 'develop' into SC-247-endpoint-do-dashboardu-studenta

kwdrt commented:
Przy próbie rozwiązania merge conflictów nie mogę postawić backu, sypie się przy budowie - @barto14753 możesz rzucić na to okiem?

barto14753 committed:
[SC-247] Merge remote-tracking branch 'origin/develop' into SC-247-endpoint-do-dashboardu-studenta

# Conflicts:
#	api/src/main/java/com/example/api/config/DatabaseConfig.java

kwdrt commented:
Jest git, teraz udało mi się zbudować

barto14753 committed:
[SC-247] Merge branch 'develop' into SC-247-endpoint-do-dashboardu-studenta

# Conflicts:
#	api/src/main/java/com/example/api/config/DatabaseConfig.java

barto14753 committed:
[SC-247] Fixed getMapTasks

barto14753 committed:
[SC-247] Merge remote-tracking branch 'origin/SC-247-endpoint-do-dashboardu-studenta' into SC-247-endpoint-do-dashboardu-studenta

# Conflicts:
#	api/src/main/java/com/example/api/config/DatabaseConfig.java

barto14753 committed:
[SC-247] Fixed login to student account

barto14753 commented:
Domergowałem developa i musiałem naprawić co nieco, bo wywalało widok edycji rozdziału u prowadzącego po zmianie wymagań i jeszcze nie dało się zalogować na studenta
